File: app/src/main/java/com/lusle/android/soon/View/Main/MainActivity.java
package com.lusle.android.soon.View.Main;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.firebase.FirebaseApp;
import com.google.firebase.iid.FirebaseInstanceId;
import com.google.firebase.messaging.FirebaseMessaging;
import com.lusle.android.soon.View.BaseActivity;
import com.lusle.android.soon.View.Search.SearchActivity;
import com.lusle.android.soon.View.Main.Company.CompanyFragment;
import com.lusle.android.soon.View.Main.Date.DateFragment;
import com.lusle.android.soon.View.Main.Genre.GenreFragment;
import com.lusle.android.soon.View.Main.Setting.SettingFragment;
import com.lusle.android.soon.View.Main.ThisMonthMovie.ThisMonthMovieFragment;
import com.lusle.android.soon.R;
import com.pranavpandey.android.dynamic.toasts.DynamicToast;

import androidx.core.app.ActivityCompat;
import androidx.core.app.ActivityOptionsCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import butterknife.ButterKnife;

public class MainActivity extends BaseActivity {

    private BottomNavigationView bottomNavigationView;
    private FloatingActionButton mSearchFab;
    private AdView adView;

    private final long FINISH_INTERVAL_TIME = 2000;
    private long backPressedTime = 0;

    final private int requestCode = 666;

    private String TAG = MainActivity.class.getSimpleName();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        init();
    }

    private void init() {
        //MobileAds.initialize(this, "ca-app-pub-2329923322434251~4419072683");
        FirebaseApp.initializeApp(this);
        FirebaseMessaging.getInstance().subscribeToTopic("all");
        FirebaseInstanceId.getInstance().getInstanceId()
                .addOnCompleteListener(task -> {
                    if (!task.isSuccessful()) {
                        Log.w("FCM Token", "getInstanceId failed", task.getException());
                        return;
                    }

                    // Get new Instance ID token
                    String token = task.getResult().getToken();

                    // Log and toast
                    Log.d("FCM Token", token);
                });

        adView = findViewById(R.id.adView);
        adView.setAdListener(new AdListener() {
            @Override
            public void onAdClosed() {
                super.onAdClosed();
                Log.d(TAG, "onAdClosed: ");
            }

            @Override
            public void onAdFailedToLoad(int i) {
                super.onAdFailedToLoad(i);
                Log.d(TAG, "onAdFailedToLoad: ");
            }

            @Override
            public void onAdLeftApplication() {
                super.onAdLeftApplication();
                Log.d(TAG, "onAdLeftApplication: ");
            }

            @Override
            public void onAdOpened() {
                super.onAdOpened();
                Log.d(TAG, "onAdOpened: ");
            }

            @Override
            public void onAdLoaded() {
                super.onAdLoaded();
                Log.d(TAG, "onAdLoaded: ");
            }

            @Override
            public void onAdClicked() {
                super.onAdClicked();
                Log.d(TAG, "onAdClicked: ");
            }

            @Override
            public void onAdImpression() {
                super.onAdImpression();
                Log.d(TAG, "onAdImpression: ");
            }
        });
        AdRequest adRequest = new AdRequest.Builder().build();
        adView.loadAd(adRequest);

        mSearchFab = findViewById(R.id.floatingActionButton);
        mSearchFab.setOnClickListener(this::presentActivity);

        bottomNavigationView = findViewById(R.id.navigation);
        bottomNavigationView.setOnNavigationItemSelectedListener(item -> {
            Fragment selectedFragment = null;
            switch (item.getItemId()) {
                case R.id.home:
                    selectedFragment = ThisMonthMovieFragment.newInstance();
                    break;
                case R.id.company:
                    selectedFragment = CompanyFragment.newInstance();
                    break;
                case R.id.genre:
                    selectedFragment = GenreFragment.newInstance();
                    break;
                case R.id.date:
                    selectedFragment = DateFragment.newInstance();
                    break;
                case R.id.settings:
                    selectedFragment = SettingFragment.newInstance();
                    break;
            }
            FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
            transaction.replace(R.id.frame_layout, selectedFragment);
            transaction.commit();
            return true;
        });

        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.replace(R.id.frame_layout, ThisMonthMovieFragment.newInstance());
        transaction.commit();

        bottomNavigationView.getMenu().getItem(0).setChecked(true);
    }

    @Override
    public void onBackPressed() {
        long tempTime = System.currentTimeMillis();
        long intervalTime = tempTime - backPressedTime;

        if (0 <= intervalTime && FINISH_INTERVAL_TIME >= intervalTime) {
            super.onBackPressed();
        } else {
            backPressedTime = tempTime;
            DynamicToast.make(this, "종료하고 싶다면 한번 더 눌러주세요").show();
        }
    }

    public void presentActivity(View view) {
        ActivityOptionsCompat options = ActivityOptionsCompat.makeSceneTransitionAnimation(this, view, "transition");
        int revealX = (int) (view.getX() + view.getWidth() / 2);
        int revealY = (int) (view.getY() + view.getHeight() / 2);

        Intent intent = new Intent(this, SearchActivity.class);
        intent.putExtra(SearchActivity.EXTRA_CIRCULAR_REVEAL_X, revealX);
        intent.putExtra(SearchActivity.EXTRA_CIRCULAR_REVEAL_Y, revealY);

        ActivityCompat.startActivityForResult(this, intent, requestCode, options.toBundle());
    }
}

File: app/src/main/java/com/lusle/android/soon/View/Main/MainActivity.kt
package com.lusle.android.soon.View.Main

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.core.app.ActivityCompat
import androidx.core.app.ActivityOptionsCompat
import androidx.lifecycle.LiveData
import androidx.navigation.NavController
import com.example.android.navigationadvancedsample.setupWithNavController
import com.google.android.gms.tasks.Task
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.firebase.FirebaseApp
import com.google.firebase.iid.FirebaseInstanceId
import com.google.firebase.iid.InstanceIdResult
import com.google.firebase.messaging.FirebaseMessaging
import com.lusle.android.soon.R
import com.lusle.android.soon.View.BaseActivity
import com.lusle.android.soon.View.Search.SearchActivity

class MainActivity : BaseActivity() {
    private lateinit var bottomNavigationView: BottomNavigationView
    private lateinit var searchFab: FloatingActionButton
    private val requestCode = 666
    private var currentNavController: LiveData<NavController>? = null
    private val TAG = MainActivity::class.java.simpleName
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        if(savedInstanceState == null){
            setUpBottomNavigationBar()
        }
        init()
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        setUpBottomNavigationBar()
    }

    private fun setUpBottomNavigationBar() {
        bottomNavigationView = findViewById(R.id.navigation_view)
        bottomNavigationView.menu.findItem(R.id.blank).isEnabled = false
        val navGraphIds = listOf(R.navigation.navigation_home, R.navigation.navigation_company, R.navigation.navigation_genre, R.navigation.navigation_settings)

        val controller = bottomNavigationView.setupWithNavController(
                navGraphIds, supportFragmentManager, R.id.nav_host_fragment, intent
        )

        // Whenever the selected controller changes, setup the action bar.
        /*
        controller.observe(this, Observer { navController -> setupActionBarWithNavController(navController) })
        */
        //액션바 관련 코드 (액션바를 사용하지 않기 때문에 주석처리)

        currentNavController = controller

        bottomNavigationView.setOnNavigationItemReselectedListener {
            when(it.itemId){
                R.id.home -> {

                }
                R.id.company -> {}
                R.id.genre -> {}
            }
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        return currentNavController?.value?.navigateUp() ?: false
    }

    private fun init() {
        //MobileAds.initialize(this, "ca-app-pub-2329923322434251~4419072683");
        FirebaseApp.initializeApp(this)
        FirebaseMessaging.getInstance().subscribeToTopic("all")
        FirebaseInstanceId.getInstance().instanceId
                .addOnCompleteListener { task: Task<InstanceIdResult> ->
                    if (!task.isSuccessful) {
                        Log.w("FCM Token", "getInstanceId failed", task.exception)
                        return@addOnCompleteListener
                    }

                    // Get new Instance ID token
                    val token = task.result.token

                    // Log and toast
                    Log.d("FCM Token", token)
                }
        searchFab = findViewById(R.id.floatingActionButton)
        searchFab.setOnClickListener(View.OnClickListener { view: View -> presentActivity(view) })
    }

    private fun presentActivity(view: View) {
        val options = ActivityOptionsCompat.makeSceneTransitionAnimation(this, view, "transition")
        val revealX = (view.x + view.width / 2).toInt()
        val revealY = (view.y + view.height / 2).toInt()
        val intent = Intent(this, SearchActivity::class.java)
        intent.putExtra(SearchActivity.EXTRA_CIRCULAR_REVEAL_X, revealX)
        intent.putExtra(SearchActivity.EXTRA_CIRCULAR_REVEAL_Y, revealY)
        ActivityCompat.startActivityForResult(this, intent, requestCode, options.toBundle())
    }
}